### DISCUSSION TITLE
Get Permissions when User is queried

### DISCUSSION THREAD

**Discussion by ClementineZou** [prisma/prisma#14205] [Q&A] [answered]
Hi, I have two model that called `UserBasic` and `AuthoRolePermission`. I want to get all the permissions that the role binded to user has when I query a user, how can I make it?
```prisma
model UserBasic {
  id String @unique @default(uuid())

  email        String  @unique
  nickName     String
  password     String
  verifyMethod String  @default("['email']")
  region       String  @default("Hong Kong")
  role         String  @default("['user']")
  prefix       String?
  displayName  String?
  avatarURL    String?
  otpSecret    String?

  signinMethod        UserSigninMethod?
  thirdParty          UserThirdParty?
  contact             UserContact?
  detail              UserDetail?
  bans                UserBan[]
  logs                UserLog[]
  forgetPasswordCodes UserForgetPasswordCode[]
  addresses           UserAddress[]
  additionPermissions AuthoAdditionPermission[]
  denyPermissions     AuthoDenyPermission[]
}
```

```prisma
model AuthoRolePermission {
  id Int @id @default(autoincrement())

  role       String
  action     String
  objectName String
}
```

**Comment by TasinIshmam** [comment, marked as answer]
Hey @AuroraAndersson 👋  

The way you have it set up right now, you'll have to do a separate query on the `AuthoRolePermission` table. But it might be possible to do this in a single query if you were willing to change your schema a bit and change `role` from a `string` to a relation to the `AuthoRolePermission` table. 

Is it alright to use two queries for this, or would you be willing to change your schema? I can elaborate further based on your preference.

**Reply by ClementineZou** [reply]
Hi! Thanks for your replying. I would like to change my schema to do a single query because I am just started. Could you please show me how to make it? Thank you

**Reply by ClementineZou** [reply]
For example, A user `linda` is an admin, so her role should be  ["user","admin"] which means she will have the permissions both `user role` and `admin role` have. In `AuthoRolePermission`, we will have many records binded to one role to define what permissions the role has. And then, I want to get the permissions just as I said before.

**Reply by TasinIshmam** [reply]
To clarify, the permissions themselves are stored in the `AuthoRolePermission` records, right?  @AuroraAndersson

**Reply by ClementineZou** [reply]
Yes, you're right

**Reply by TasinIshmam** [reply]
I would change your schema like this: 

```prisma
model UserBasic {
  id           String                @id @unique @default(uuid())
  email        String                @unique
  nickName     String
  password     String
  verifyMethod String                @default("['email']")
  region       String                @default("Hong Kong")
  roles        AuthoRolePermission[]   //  a user can have many roles 
  prefix       String?
  displayName  String?
  avatarURL    String?
  otpSecret    String?

  signinMethod        UserSigninMethod?
  thirdParty          UserThirdParty?
  contact             UserContact?
  detail              UserDetail?
  bans                UserBan[]
  logs                UserLog[]
  forgetPasswordCodes UserForgetPasswordCode[]
  addresses           UserAddress[]
  additionPermissions AuthoAdditionPermission[]
  denyPermissions     AuthoDenyPermission[]

}

model AuthoRolePermission {
  role       String      @id @unique   // The role should be unique. You like don't need to have a separate `id` field either, the role itself can serve as the primary key. 
  action     String
  objectName String
  users      UserBasic[]   // a role can be assigned to many users
}

```

Note that I'm assuming the `AuthoRolePermission.role` field stores the name of a particular role. I suspect you don't need a separate `id` field in `AuthoRolePermission`, the `role` field itself can server that purpose. In case this assumption is wrong, feel free to let me know. 

You can now do queries like this: 

```ts
// get a user and their roles
const user = await prisma.userBasic.findUnique(
  {
    where: {
      id: "57208875-047e-4142-9a95-4c12c633387d"
    },
    include: {
      roles: true,
    }
  }
)

// update a user to add a new role
const updateUserData = await prisma.userBasic.update({
  where: {
    id: "57208875-047e-4142-9a95-4c12c633387d"
  },
  data: {
    roles: {
      connect: {
        role: "admin"
      }
    }
  }
})

```

The only drawback of this approach is you lose the ability to create `default` roles for the user. This is something you'll have to handle inside your application code so that the user is automatically assigned a default role, in case one is not specified.  

Let me know if everything is clear!

**Reply by ClementineZou** [reply]
Thank you very much! I know what you wrote. But if I use another way to store Role and Permission such as the following example records, what should I do to the schema?
```
{role: "user",action:"update",objectName:"address"}
{role: "user",action:"create",objectName:"post"}
```

**Reply by TasinIshmam** [reply]
Oh I see. In that case I think it would be best to do two separate queries: 

1. Query `UserBasic` to find a User and get the `role` name. 
2. Do a `findMany` on the `AuthoRolePermission` table to find all permissions for that certain role.

**Reply by ClementineZou** [reply]
Thanks for your reply! Have a good day!

**Reply by TasinIshmam** [reply]
One additional point, @AuroraAndersson  

The `role` field can be a Scalar List `String[]` instead of a `String`, so you don't have to parse the roles manually from a string. 

More information in the [prisma docs](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-scalar-lists-arrays).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
